RFCs/FS-1135-random-functions-for-collections.md

```diff
@@ -0,0 +1,206 @@
+```
+[ArgumentNullException](https://learn.microsoft.com/en-us/dotnet/api/system.argumentnullexception) should be raised if collection is null
+
+[ArgumentException](https://learn.microsoft.com/en-us/dotnet/api/system.argumentoutofrangeexception) should be raised if N is negative
+
+[ArgumentException](https://learn.microsoft.com/en-us/dotnet/api/system.argumentexception) should be raised if collection is empty
+
+Example:
+```fsharp
+let allPlayers = [ "Alice"; "Bob"; "Charlie"; "Dave" ]
+let round1Order = allPlayers |> List.randomChoices 3 // ["Bob", "Dave", "Bob"]
+```
+
+### Sample
+
+Sample should select N elements from input collection in random order, once element is taken it won't be selected again. N can't be greater than collection length
+
+Two functions should be added to each module.
+
+```fsharp
+// Array module
+val randomSample: count:int -> array:'T[] -> 'T[]
+val randomSampleWith: random:Random -> count:int -> array:'T[] -> 'T[]
+// List module
+val randomSample: count:int -> list:'T list -> 'T list
+val randomSampleWith: random:Random -> count:int -> list:'T list -> 'T list
+// Seq module
+val randomSample: count:int -> source:'T seq -> 'T seq
+val randomSampleWith: random:Random -> count:int -> source:'T seq -> 'T seq
+```
+[ArgumentNullException](https://learn.microsoft.com/en-us/dotnet/api/system.argumentnullexception) should be raised if collection is null
+
+[ArgumentException](https://learn.microsoft.com/en-us/dotnet/api/system.argumentoutofrangeexception) should be raised if N is greater than collection length or is negative
+
+[ArgumentException](https://learn.microsoft.com/en-us/dotnet/api/system.argumentexception) should be raised if collection is empty
+
+Example:
+```fsharp
+let allPlayers = [ "Alice"; "Bob"; "Charlie"; "Dave" ]
+let round1Order = allPlayers |> List.randomSample 3 // ["Charlie", "Dave", "Alice"]
+```
+
+# Drawbacks
+
+System.Random interface has added some new methods in [.NET 8](https://learn.microsoft.com/en-us/dotnet/core/whats-new/dotnet-8#methods-for-working-with-randomness), where naming is a bit different. More new methods can eventually be added in future .NET versions.
+
+# Alternatives
+
+Use online snippets, or provide a nuget package.
+
+# Compatibility
+
+Please address all necessary compatibility questions:
+
+* Is this a breaking change? **No, unless user defined those extensions themselves**
+* What happens when previous versions of the F# compiler encounter this design addition as source code? **Should be fine**
+* What happens when previous versions of the F# compiler encounter this design addition in compiled binaries? **Should be fine**
+* If this is a change or extension to FSharp.Core, what happens when previous versions of the F# compiler encounter this construct? **Will work as usual**
+
+# Pragmatics
+
+## Diagnostics
+
+Please list the reasonable expectations for diagnostics for misuse of this feature. **I don't see a way to misuse it**
```

Comment:
Not sure that goes in here, but any random implementation can be valid even if it always returns the exact same value. In that case, the same item would always be returned. But since this is essentially true of anything that anybody can write in their own code, it's probably not worth mentioning.

Reply:
I don't know what to write here either :) So leaved it as is

Reply:
Hey @Lanayx, it appears we both forgot about this line. We shouldn't leave the remnants of the template hanging around (I know, it happens, but we shouldn't ;) ).

This is my suggestion:

```suggestion
There are no known diagnostics on any abuse or misuse of this feature.
```